typing/types.ml

```diff
@@ -441,45 +453,49 @@ type constructor_description =
     cstr_uid: Uid.t;
    }
 
-and constructor_tag =
-    Cstr_constant of int                (* Constant constructor (an int) *)
-  | Cstr_block of int                   (* Regular constructor (a block) *)
-  | Cstr_unboxed                        (* Constructor of an unboxed type *)
-  | Cstr_extension of Path.t * bool     (* Extension constructor
-                                           true if a constant false if a block*)
-
 let equal_tag t1 t2 =
   match (t1, t2) with
-  | Cstr_constant i1, Cstr_constant i2 -> i2 = i1
-  | Cstr_block i1, Cstr_block i2 -> i2 = i1
-  | Cstr_unboxed, Cstr_unboxed -> true
-  | Cstr_extension (path1, b1), Cstr_extension (path2, b2) ->
-      Path.same path1 path2 && b1 = b2
-  | (Cstr_constant _|Cstr_block _|Cstr_unboxed|Cstr_extension _), _ -> false
+  | Ordinary {src_index=i1}, Ordinary {src_index=i2} -> i2 = i1
+  | Extension (path1,_), Extension (path2,_) -> Path.same path1 path2
+  | (Ordinary _ | Extension _), _ -> false
 
 let may_equal_constr c1 c2 =
   c1.cstr_arity = c2.cstr_arity
   && (match c1.cstr_tag,c2.cstr_tag with
-     | Cstr_extension _,Cstr_extension _ ->
+     | Extension _, Extension _ ->
          (* extension constructors may be rebindings of each other *)
          true
      | tag1, tag2 ->
          equal_tag tag1 tag2)
 
+let kind_abstract ~layout = Type_abstract { layout }
+let kind_abstract_value = kind_abstract ~layout:(Sort Value)
+let kind_abstract_immediate = kind_abstract ~layout:Immediate
+let kind_abstract_any = kind_abstract ~layout:Any
+
+let decl_is_abstract decl =
+  match decl.type_kind with
+  | Type_abstract _ -> true
+  | Type_record _ | Type_variant _ | Type_open -> false
+
 type label_description =
   { lbl_name: string;                   (* Short name *)
     lbl_res: type_expr;                 (* Type of the result *)
     lbl_arg: type_expr;                 (* Type of the argument *)
     lbl_mut: mutable_flag;              (* Is this a mutable field? *)
-    lbl_global: global_flag;        (* Is this a global field? *)
+    lbl_global: global_flag;            (* Is this a global field? *)
```

Comment:
I agree the inconsistency about when such comments appear in the ml is annoying.  I think, as this is orthogonal to unboxed types, it would make our lives easier in future merges to make this improvement in one commit on all such types.  For the same reason, I've undone my change to the spacing.